plutus-core/plutus-ir/src/PlutusIR/Analysis/Builtins.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE KindSignatures   #-}
+{-# LANGUAGE RankNTypes       #-}
+{-# LANGUAGE TemplateHaskell  #-}
+{-# LANGUAGE TypeApplications #-}
+module PlutusIR.Analysis.Builtins where
+
+import Control.Lens hiding (parts)
+import Data.Kind
+import Data.Map qualified as Map
+import Data.Proxy
+import PlutusCore.Builtin
+import PlutusCore.Builtin qualified as PLC
+import PlutusCore.Default
+import PlutusIR.Arity
+import PlutusIR.Contexts
+import PlutusIR.Core.Type hiding (Type)
+import PlutusPrelude (Default (..))
+
+-- | The information we need to work with a builtin that is like a datatype matcher.
+data BuiltinMatcherLike uni fun = BuiltinMatcherLike
+  { _bmlSplitMatchContext :: forall tyname name a .
+      (AppContext tyname name uni fun a ->
+       Maybe (SplitMatchContext tyname name uni fun a, [Arity],
+              SplitMatchContext tyname name uni fun a -> AppContext tyname name uni fun a))
+  }
+makeLenses ''BuiltinMatcherLike
+
+data BuiltinsInfo (uni :: Type -> Type) fun = BuiltinsInfo
+  { _biSemanticsVariant :: PLC.BuiltinSemanticsVariant fun
+  , _biMatcherLike      :: Map.Map fun (BuiltinMatcherLike uni fun)
+  }
+
+makeLenses ''BuiltinsInfo
+
+instance (Ord fun, Default (BuiltinSemanticsVariant fun)) => Default (BuiltinsInfo uni fun) where
+  def = BuiltinsInfo def mempty
+
+-- | Get the 'Arity' from a 'TypeScheme'.
+typeSchemeArity :: TypeScheme val args res -> Arity
+typeSchemeArity TypeSchemeResult{}    = []
+typeSchemeArity (TypeSchemeArrow sch) = TermParam : typeSchemeArity sch
+typeSchemeArity (TypeSchemeAll _ sch) = TypeParam : typeSchemeArity sch
+
+-- | Get the arity of a builtin function.
+builtinArity
+    :: forall uni fun
+    . ToBuiltinMeaning uni fun
+    => BuiltinsInfo uni fun
+    -> fun
+    -> Arity
+builtinArity binfo fun =
+  let semvar = binfo ^. biSemanticsVariant
+  in case toBuiltinMeaning @uni @fun @(Term TyName Name uni fun ()) semvar fun of
+        BuiltinMeaning sch _ _ -> typeSchemeArity sch
+
+-- | Get the arity of a builtin function.
+builtinArity'
```

Comment:
One just takes the `BuiltinSemanticsVariant`, one takes a `BuiltinsInfo`. I meant to define one in terms of the other... I wanted `builtinArity'` because it's more convenient to use in `defaultUniMatcherLike` below :thinking: